--- a/src/rdiff_backup/connection.py
+++ b/src/rdiff_backup/connection.py
@@ -31,6 +31,7 @@ import tempfile  # noqa: F401
 import types  # noqa: F401
 import time
 import subprocess
+import errno
 
 # The following EA and ACL modules may be used if available
 try:  # compat200
@@ -430,7 +431,12 @@ class PipeConnection(LowLevelPipeConnection):
             self._put(arg, req_num)
         result = self._get_response(req_num)
         self.unused_request_numbers.add(req_num)
-        if isinstance(result, Exception):
+        if isinstance(result, OSError) and getattr(result, "errno_str", False):
+            # OSError code are specific to each platform.
+            # Let convert the errno to current platform.
+            result.errno = getattr(errno, result.errno_str, -result.errno)
+            raise result
+        elif isinstance(result, Exception):
             raise result
         elif isinstance(result, SystemExit):
             raise result
@@ -522,7 +529,12 @@ class PipeConnection(LowLevelPipeConnection):
                         ex=sys.exc_info()[1], ty=sys.exc_info()[0],
                         tb="".join(traceback.format_tb(sys.exc_info()[2]))),
                     log.INFO)
-        return sys.exc_info()[1]
+        result = sys.exc_info()[1]
+        # OSError code are specific to the platform. Send back code as string.
+        if isinstance(result, OSError) and result.errno:
+            result.errno_str = errno.errorcode.get(result.errno, "EUNKWN")
+            result.strerror = f'[original: Errno {result.errno} {result.errno_str}] {result.strerror}'
+        return result
 
     def _get_new_req_num(self):
         """Allot a new request number and return it"""
--- a/src/rdiff_backup/iterfile.py
+++ b/src/rdiff_backup/iterfile.py
@@ -18,6 +18,7 @@
 # 02110-1301, USA
 """Convert an iterator to a file object and vice-versa"""
 
+import errno
 import pickle
 import array
 from . import Globals, robust, rpath
@@ -272,6 +273,10 @@ class FileWrappingIter:
 
     def _read_error_handler(self, exc, blocksize):
         """Log error when reading from file"""
+        # OSError code are specific to the platform. Send back code as string.
+        if isinstance(exc, OSError) and exc.errno:
+            exc.errno_str = errno.errorcode.get(exc.errno, "EUNKWN")
+            exc.strerror = f'[original: Errno {exc.errno} {exc.errno_str}] {exc.strerror}'
         self.last_exception = exc
         return None
 
@@ -489,6 +494,10 @@ class ErrorFile:
 
     def __init__(self, exc):
         """Initialize new ErrorFile.  exc is the exception to raise on read"""
+        if isinstance(exc, OSError) and getattr(exc, "errno_str", False):
+            # OSError code are specific to each platform.
+            # Let convert the errno to current platform.
+            exc.errno = getattr(errno, exc.errno_str, -exc.errno)
         self.exc = exc
 
     def read(self, lines=-1):